### PR TITLE
Making Blue-Green Deployments its own topic

### PR DESCRIPTION
**Pull request by ahardin-rh**
After some discussion with @adellape, we agree it may be best to make the Blue-Green Deployments content its own topic under Upgrading.

Original work is here:
https://github.com/openshift/openshift-docs/pull/2287
https://github.com/openshift/openshift-docs/pull/2511

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@brenton PTAL
@adellape please peer review

File: install_config/upgrading/blue_green_deployments.adoc [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+:toc: macro
+:toc-title:
+:prewrap!:
+
+toc::[]
+
+== Overview
+
+Blue-green deployments are a proven approach to reducing downtime caused while
+updating an environment. This is done by creating a parallel environment on
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe s/updating/upgrading/, to keep in line w/ the subdir name.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+which the new deployment can be installed. If a problem is detected and after
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Should there be a comma after "detected"? Or drop the word "and"?

File: install_config/upgrading/blue_green_deployments.adoc [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+the new deployment is verified, traffic can be switched over with the option to
+rollback.
+
+While blue-green is a valid strategy for deploying just about any software,
+there are always trade-offs to be made. Not all environments have the same
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe:

s/to be made//

File: install_config/upgrading/blue_green_deployments.adoc [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+uptime requirements or the resources to properly perform blue-green deployments.
+In an {product-title} environment the most suitable candidate for blue-green
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: "environment,"?

File: install_config/upgrading/blue_green_deployments.adoc [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+deployments are the Nodes. Uptime is most important for these workloads and the
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/Nodes/nodes/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+additional complexity of blue-green deployments can be justified. Often the main
+challenge is having the excess capacity to facilitate such an approach.
+
+ifdef::openshift-enterprise[]
+Another lesser challenge is that the administrator will have to temporarily
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/will have to/must/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+share the entitlements between the blue-green deployments or provide access to
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe be more specific here:

s/entitlements/Red Hat software entitlements/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+the installation content by means of a system such as Red Hat Satellite. This
+can be accomplished by sharing the consumer ID from the previous host.
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe end this with a colon and turn the next steps into a numbered list?

File: install_config/upgrading/blue_green_deployments.adoc [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+On the new host:
+----
+# subscription-manager register --consumerid=6699375b-06db-48c4-941e-689efd6ce3aa
+----
+
+[IMPORTANT]
+====
+After a successful deployment, remember to unregister the old host with
+`*subscription-manager clean*` to prevent the environment from being out of
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/`*subscription-manager clean*`/`subscription-manager clean`/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+compliance.
+====
+
+endif::openshift-enterprise[]
+
+[[manual-upgrades-preparing-for-upgrade]]
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Probably update these unique IDs that start with `[[manual-upgrades` (there are more below) to start with `[[blue-green` or something.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+== Preparing for Upgrade
+
+xref:../../install_config/upgrading/manual_upgrades.adoc#upgrading-masters[After
+the master and etcd servers have been upgraded], you need to ensure that your
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/need to/must/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+current production nodes are labeled either blue or green. In this example, the
+current installation will be blue and the new environment will be green.
+
+----
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe add a line before the code block like:

"On each production node in your current installation:"

File: install_config/upgrading/blue_green_deployments.adoc [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+$ oc label --all nodes color=blue
+----
+
+In the case of nodes requiring the uptime guarantees of a blue-green deployment,
+the `-l` flag can be used to match a subset of the environment using a selector.
+
+Next, xref:../install/advanced_install.adoc#adding-nodes-advanced[the green
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Just FYI, @bfallonf has a PR (https://github.com/openshift/openshift-docs/pull/2570) that is proposing to move the "adding hosts" content to its own new topic. So that would affect this.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+environment] should be created for any nodes that are to be replaced. Ansible
+can apply the `*color=green*` using the `*openshift_node_labels*` variable for
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]:
```
can apply the `*color=green*` label using...
```

File: install_config/upgrading/blue_green_deployments.adoc [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+each node.
+
+In order to delay workload scheduling until the nodes are
+xref:../../dev_guide/application_health.adoc#dev-guide-application-health[healthy],
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: That dev_guide topic is about application health, whereas this sentence is referring to node health, like when you run a "oc get nodes" to see if they're Ready or NotReady. See the "Node Conditions" table here https://docs.openshift.com/enterprise/latest/admin_guide/manage_nodes.html#listing-nodes and the general stuff about node health here https://docs.openshift.com/enterprise/latest/architecture/infrastructure_components/kubernetes_infrastructure.html#node.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+be sure to set the `*openshift_schedulable=false*` variable.
+
+Once the green nodes are in *Ready* state, they can be made schedulable. Then,
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/Once/After/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+blue nodes are disabled so that no new pods are run on them.
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: s/./:/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+still available so no capacity is lost.
+
+[[manual-upgrades-warming-the-new-nodes]]
+== Warming the New Nodes
+
+In order for pods to be migrated from the blue environment to the green, the
+images must be pulled. Network latency and load on the registry can cause delays
+if there is not sufficient capacity built in to the environment. Often, the best
+way to minimize impact to the running system is to trigger new pod deployments that
+will land on the new nodes. Accomplish this by importing new *ImageStreams*.
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Can just be simply:

s/*ImageStreams*/image stream/

File: install_config/upgrading/blue_green_deployments.adoc [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+A major release of {product-title} is the motivation for a
+blue-green deployment. At that time, new *ImageStreams* become
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Same for any other instances of `*ImageStreams*`.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+available for users of Source-to-Image. Upon import, any builds or deployments
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Source-to-Image (S2I)

File: install_config/upgrading/blue_green_deployments.adoc [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+configured with *ImageChangeTriggers* are automatically created.
+
+It is important to realize that this process can trigger a large number of
+builds. The good news is that the builds are performed on the green nodes and,
+therefore, do not impact any traffic on the blue deployment.
+
+Monitor build progress:
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe: "To monitor build progress:"

Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Or even "To monitor build progress across all namespaces (projects) in the cluster:"

File: install_config/upgrading/blue_green_deployments.adoc [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+== Node Evacuation
+
+For larger deployments, it is possible to have other labels that help
+determine how evacuation can be coordinated. The most conservative approach
+for avoiding downtime is to evacuate one node at a time. If services are
+composed of pods using zone anti-affinity, then an entire zone can be
+evacuated at once. It is important to ensure that the storage volumes used are
+available in the new zone as this detail can vary among cloud providers.
+
+In OpenShift Enterprise 3.2 and later, a node evacuation is triggered whenever
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe split this using ifdefs and {product-title}s here so that it comes out saying "OpenShift Enterprise 3.2" for enterprise branches, and "OpenShift Origin 1.2" for origin.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+the service is stopped. Achieve manual evacuation and deletion of all blue nodes
+at once by:
+
+----
+# oadm manage-node --selector=color=blue --evacuate
+# oc delete node --selector=color=blue
+----
+
+[[manual-upgrades-summary]]
+== Summary
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Confused by the placement of this summary at the end. At first I thought it was trying to be like an "In conclusion..." sort of wrap up, but it seems to have a lot of repeating language from the the intro to the topic.

**Comment by adellape** [comment]
@ahardin-rh Done with my comments!

File: install_config/upgrading/blue_green_deployments.adoc (hunk 23 above)
Comment (ahardin-rh) [inline install_config/upgrading/blue_green_deployments.adoc]: Good point. Removed the summary and added some bits to the Overview.

**Comment by ahardin-rh** [comment]
@adellape appreciate the eyeballs! :eyes:

**Comment by ahardin-rh** [comment]
@brenton how much downtime can a user expect when upgrading and how much time would this save? Thinking of adding a line in the Upgrading overview in a separate PR so users know what to expect. Thanks!

**Comment by brenton** [comment]
@ahardin-rh, it's really a matter of whether or not things go as planned or if there are unforeseen problems.  Restarting hosts for the upgrade is a fairly quick operation.  However, in mission critical environments no one relies on things working perfectly every time so blue-green deployments are an approach to manage the risk.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+deployments are the nodes. All user processes run on these systems and even
+critical pieces of {product-title} infrastructure are self-hosted there. Uptime
+is most important for these workloads and the additional complexity of
+blue-green deployments can be justified. The exact implementation of this
+approach varies based on your requirements. Often the main challenge is having the
+excess capacity to facilitate such an approach.
+
+ifdef::openshift-enterprise[]
+Another lesser challenge is that the administrator must temporarily share the
+Red Hat software entitlements between the blue-green deployments or provide
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Recommended to @ahardin-rh to consider breaking this up, leaving the "lesser challenge" part here and moving the actual steps into "Preparing for Upgrade" below.

Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Also recommended adding some info at the end of the Overview to the effect of: "This topic serves as an alternative node upgrade method to the approach in Manual Upgrades <link to that topic's Upgrading Nodes>". So that if someone comes to this topic cold turkey, there's some context that this is specifically for upgrading nodes right now and that there are other steps they'd have to do first before this topic.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+
+----
+$ oc label --all nodes color=blue
+----
+
+In the case of nodes requiring the uptime guarantees of a blue-green deployment,
+the `-l` flag can be used to match a subset of the environment using a selector.
+
+Next, xref:../install/advanced_install.adoc#adding-nodes-advanced[the green
+environment] should be created for any nodes that are to be replaced. Ansible
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Maybe rephrase this like "Next, create the new green environment for any nodes that are to be replaced."

Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: "... by adding an equal number of new nodes to the existing cluster"? And maybe more clearly show somehow that we're linking to the "Adding New Nodes" section? I say that because I wasn't sure what I was being told to here until I clicked on "the green environment".

File: install_config/upgrading/blue_green_deployments.adoc [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+
+A common practice is to scale the registry and router pods until they are
+migrated to the green nodes. For these pods, a _canary_ deployment approach is
+commonly used. Scaling them up will make them immediately active on the new
+nodes. Pointing the deployment configuration to the new image initiates a
+rolling update. However, because of node anti-affinity, and the fact that the
+blue nodes are still unschedulable, the deployments to the old nodes will fail.
+At this point, the registry and router deployments can be scaled down to the
+original number of pods. At any given point, the original number of pods is
+still available so no capacity is lost.
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: Is there a way the tasks in this "Preparing for Upgrade" section can be broken up into numbered steps? Seems like that would be easier to grok at a glance.

File: install_config/upgrading/blue_green_deployments.adoc [hunk 27]
```diff
@@ -0,0 +1,166 @@
...
+In order for pods to be migrated from the blue environment to the green, the
+images must be pulled. Network latency and load on the registry can cause delays
+if there is not sufficient capacity built in to the environment. Often, the best
+way to minimize impact to the running system is to trigger new pod deployments that
+will land on the new nodes. Accomplish this by importing new image streams.
+
+A major release of {product-title} is the motivation for a blue-green
+deployment. At that time, new image streams become available for users of
+Source-to-Image (S2I). Upon import, any builds or deployments configured with
+*ImageChangeTriggers* are automatically created.
```
Comment (adellape) [inline install_config/upgrading/blue_green_deployments.adoc]: @brenton Should they be instructed at this point to refer to these topics for steps?

https://docs.openshift.com/enterprise/latest/install_config/upgrading/manual_upgrades.html#updating-the-default-image-streams-and-templates
https://docs.openshift.com/enterprise/latest/install_config/upgrading/manual_upgrades.html#importing-the-latest-images

Comment (brenton) [inline install_config/upgrading/blue_green_deployments.adoc]: That sounds like a good idea.

**Comment by ahardin-rh** [comment]
@adellape @brenton Thanks! This is updated. Good to go?

**Comment by adellape** [comment]
@ahardin-rh :+1:

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../install_config/upgrading/blue_green_deployments.adoc#upgrading-blue-green-deployments[Upgrading -> Blue-Green Deployments]
|Added new topic.
%
